File: main/java/kuro9/mahjongbot/UserGameData.java
package kuro9.mahjongbot;

import java.io.Serializable;
import java.util.Arrays;

public class UserGameData implements Serializable {
    public final String name;
    public double total_uma;
    public int game_count;
    public int[] rank_count;
    public double[] rank_pp;
    public double avg_rank;
    public double avg_uma;

    public UserGameData(String name) {
        this.name = name;
        rank_count = new int[]{0, 0, 0, 0, 0};
        rank_pp = new double[]{0, 0, 0, 0, 0};
        total_uma = 0;
        game_count = 0;
        avg_rank = 0;
        avg_uma = 0;
    }

    /**
     * 게임 결과를 현재 객체에 더합니다.
     *
     * @param score 점수
     * @param rank  순위(범위 : [1, 4])
     */
    public void addGameData(int score, int rank) {
        total_uma += (Math.round((((score - Setting.RETURN_POINT) / 1000.0) + Setting.UMA[rank - 1]) * 10) / 10.0);
        rank_count[rank - 1]++;
        if (score < 0) rank_count[4]++;
    }

    /**
     * 모든 필드를 업데이트합니다.
     */
    public void updateAllData() {
        total_uma = Math.round(total_uma * 10) / 10.0;
        game_count = Arrays.stream(rank_count).sum() - rank_count[4];

        avg_rank = 0;
        for (int i = 0; i < rank_pp.length - 1; i++) {
            rank_pp[i] = Math.round((double) rank_count[i] / (double) game_count * 10000) / 100.0;
            avg_rank += rank_count[i] * (i + 1);
        }
        rank_pp[4] = Math.round((double) rank_count[4] / (double) game_count * 10000) / 100.0;
        avg_rank = Math.round(avg_rank / game_count * 100) / 100.0;
        avg_uma = Math.round(total_uma / game_count * 10) / 10.0;
    }

    @Override
    public String toString() {
        updateAllData();
        return String.format(
                "{\"name\":\"%s\",\"total_uma\":%f,\"game_count\":%d,\"rank_count\":[%d,%d,%d,%d,%d],\"rank_pp\":[%f,%f,%f,%f,%f],\"avg_rank\":%f,\"avg_uma\":%f}",
                name,
                total_uma,
                game_count,
                rank_count[0], rank_count[1], rank_count[2], rank_count[3], rank_count[4],
                rank_pp[0], rank_pp[1], rank_pp[2], rank_pp[3], rank_pp[4],
                avg_rank,
                avg_uma
        );
    }
}

File: main/java/kuro9/mahjongbot/data/UserGameData.kt
package kuro9.mahjongbot.data

import kuro9.mahjongbot.Setting
import kuro9.mahjongbot.annotation.IntRange
import kuro9.mahjongbot.annotation.UserRes
import org.apache.commons.math3.util.Precision.round

data class UserGameData(@UserRes val id: Long) {

    /** 연산 캐싱을 위한 파라미터 */
    private var _isValid: Boolean = true
    private var _isProcessing: Boolean = false

    var _userName: String? = null
    val userName: String
        get() {
            if (_userName == null)
                _userName = Setting.JDA.retrieveUserById(id).complete().effectiveName
            return _userName!!
        }

    /** 연산 편의성 위해 10을 곱해 정수로 저장 */
    var totalUmaLong: Long = 0

    /** 실제 총 우마 값(소수 첫째 자리) */
    val totalUma: Double
        get() = totalUmaLong / 10.0

    /** 1등, 2등, 3등, 4등, 토비 */
    val rankCount: IntArray = intArrayOf(0, 0, 0, 0, 0)


    /* PRIVATE VALUE BEGIN */
    private var _gameCount: Int = 0
    private var _rankPercentage: DoubleArray = doubleArrayOf(.0, .0, .0, .0, .0)
    private var _avgRank: Double = .0
    private var _avgUma: Double = .0
    /* PRIVATE VALUE END */


    /* COMPUTED GETTER VALUE BEGIN */
    val gameCount: Int
        get() {
            checkData()
            return _gameCount
        }

    //TODO 작동하는지 확인필(소수점 반올림)
    /** 1등, 2등, 3등, 4등, 토비 */
    val rankPercentage: DoubleArray
        get() {
            checkData()
            return _rankPercentage
        }

    val avgRank: Double
        get() {
            checkData()
            return _avgRank
        }

    val avgUma: Double
        get() {
            checkData()
            return _avgUma
        }
    /* COMPUTED GETTER VALUE END */


    fun addGameData(score: Int, @IntRange(1, 4) rank: Int) {
        totalUmaLong += (((score - Setting.RETURN_POINT) / 1000.0 + Setting.UMA[rank - 1]) * 10).toLong()
        if (score < 0) rankCount[4]++
        rankCount[rank - 1]++
        _isValid = false;
    }


    private fun checkData() {
        if (!_isValid && !_isProcessing) {
            _isProcessing = true
            _gameCount = rankCount.sum() - rankCount[4]
            _rankPercentage = rankCount.map {
                round((it / gameCount.toDouble() * 100.0), 2)
            }.toDoubleArray()
            _avgRank = round(
                ((rankCount[0] + (rankCount[1] * 2) + (rankCount[2] * 3) + (rankCount[3] * 4)) / (gameCount.toDouble())),
                2
            )
            _avgUma = round(totalUma / gameCount, 2)

            _isValid = true
            _isProcessing = false
        }
    }
}